client/state/jitm/reducer.js

```diff
@@ -1,13 +1,27 @@
 /**
  * Internal dependencies
  */
-import { JITM_SET } from 'state/action-types';
-import { combineReducers, keyedReducer } from 'state/utils';
+import { JITM_SET } from 'calypso/state/action-types';
+import { combineReducers, keyedReducer } from 'calypso/state/utils';
 
 export const storeJITM = ( state = {}, { type, jitms } ) => ( type === JITM_SET ? jitms : state );
 
+const jitmCache = ( state = null, action ) => {
+	if ( action.type === JITM_SET ) {
+		if ( ! action.jitms || ! action.jitms.length ) {
+			return null;
+		}
+
+		return action.jitms[ 0 ];
+	}
+	return state;
+};
+
+const sitePathJITMCache = keyedReducer( 'siteId', jitmCache );
+
 const sitePathJITM = keyedReducer( 'keyedPath', storeJITM );
 
 export default combineReducers( {
```

Comment:
The only difference between `sitePathJITM` and `sitePathJITMCache` is that `sitePathJITMCache` doesn't save to state `null` and `[]`. Can we just keep `sitePathJITMCache` or enhance `sitePathJITM`? In my test it works, not sure though if it has any side effects.

Reply:
Can you illustrate that? Depending on what you mean, it might be one for a followup PR?

Reply:
```
/**
 * Internal dependencies
 */
import { JITM_SET } from 'calypso/state/action-types';
import { combineReducers, keyedReducer } from 'calypso/state/utils';

const storeJITM = ( state = null, action ) => {
	if ( action.type === JITM_SET ) {
		if ( ! action.jitms || ! action.jitms.length ) {
			return null;
		}

		return action.jitms[ 0 ];
	}
	return state;
};

const sitePathJITM = keyedReducer( 'keyedPath', storeJITM );

export default combineReducers( {
	sitePathJITM,
} );
```

but I just saw that in you are returning `jitms[ 0 ]` instead of `jitms` why is that?

Reply:
The key point here (see what I did there 😉 ?) is that we need x2 reducers keyed by different properties.

`sitePathJITM` keys the state by a combo of `messagePath` + `siteId`, this means JITMs are stored on a per `messagePath` basis. That's the reason why we go from `null` to JITM objects as we navigate.

I've made the cache keyed by `siteId` only. This means all JITMs from all `messagePaths` (ie: any section) are pushed into the cache. Therefore if your current `messagePath` isn't yet populated we can return the last JITM for the given `siteId` from the cache, until the network returns and populates the JITMs by `messagePath` in state.

That's why we need x2 reducers. It's all about the keys.

Reply:
Ah, yes. That explains a lot! Thanks for being thorough!